stablehlo/dialect/TypeInference.cpp

```diff
@@ -1114,20 +1103,26 @@ LogicalResult inferSliceOp(Optional<Location> location, Value operand,
   SmallVector<int64_t, 4> shape;
   shape.reserve(rank);
   SmallVector<int64_t> resultBounds;
+
+  auto inferSliceDim = [&](int64_t inputDim, int64_t start, int64_t end,
+                           int64_t stride) {
+    return inputDim == ShapedType::kDynamicSize
+               ? ShapedType::kDynamicSize
+               : static_cast<int64_t>(llvm::divideCeil(end - start, stride));
+  };
+
   for (int64_t i = 0, e = rank; i != e; i++) {
-    if (!inputBounds.empty())
-      resultBounds.push_back(
-          inferSliceDim(inputBounds[i], start[i], limit[i], strideVals[i]));
     if (hlo::isDynamicDimSize(rankedTy.getDimSize(i))) {
       shape.push_back(ShapedType::kDynamicSize);
-      continue;
+      if (inputBounds.empty()) continue;  // dynamic dim without bounds
     }
     // P3.
     if (start[i] < 0)
       return emitOptionalError(location, "negative start index ", start[i],
                                " in dimension ", i);
     // P4.
-    if (limit[i] > rankedTy.getDimSize(i))
+    if (!hlo::isDynamicDimSize(rankedTy.getDimSize(i)) &&
```

Comment:
if it is dynamic but limit > bound, then also there should be an error?

Reply:
Good point! Added code and test.